File: ios/bridging_oldArch/THEORCTAdsModule.m
//
//  THEORCTAdsModule.m
//

#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(THEORCTAdsModule, NSObject)

RCT_EXTERN_METHOD(skip:(nonnull NSNumber *)node);

RCT_EXTERN_METHOD(playing:(nonnull NSNumber *)node
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject);

RCT_EXTERN_METHOD(currentAdBreak:(nonnull NSNumber *)node
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject);

RCT_EXTERN_METHOD(currentAds:(nonnull NSNumber *)node
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject);

RCT_EXTERN_METHOD(scheduledAdBreaks:(nonnull NSNumber *)node
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject);

RCT_EXTERN_METHOD(schedule:(nonnull NSNumber *)node
                  ad:(NSDictionary)ad);

RCT_EXTERN_METHOD(daiSnapback:(nonnull NSNumber *)node
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject);

RCT_EXTERN_METHOD(daiSetSnapback:(nonnull NSNumber *)node
                  enabled:(BOOL)enabled);

RCT_EXTERN_METHOD(daiContentTimeForStreamTime:(nonnull NSNumber *)node
                  time:(nonnull NSNumber *)timeValue
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject);

RCT_EXTERN_METHOD(daiStreamTimeForContentTime:(nonnull NSNumber *)node
                  time:(nonnull NSNumber *)timeValue
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject);

RCT_EXTERN_METHOD(addFriendlyObstruction:(nonnull NSNumber *)node
                  obstruction:(NSDictionary)obstruction);

RCT_EXTERN_METHOD(removeAllFriendlyObstructions:(nonnull NSNumber *)node);

@end

File: ios/bridging_oldArch/THEORCTAdsModule.swift
//
//  THEORCTAdsModule.swift
//

@objc(THEORCTAdsModule)
class THEORCTAdsModule: NSObject, RCTBridgeModule {
    @objc var bridge: RCTBridge!
    let adsAPI = THEOplayerRCTAdsAPI()
    
    static func moduleName() -> String! {
        return "THEORCTAdsModule"
    }
    
    static func requiresMainQueueSetup() -> Bool {
        return false
    }
    
    func view(for node: NSNumber) -> THEOplayerRCTView? {
        return self.bridge.uiManager.view(forReactTag: node) as? THEOplayerRCTView
    }
    
    @objc(skip:)
    func skip(_ node: NSNumber) -> Void {
        DispatchQueue.main.async {
            self.adsAPI.skip(self.view(for: node))
        }
    }
    
    @objc(playing:resolver:rejecter:)
    func playing(_ node: NSNumber, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        DispatchQueue.main.async {
            self.adsAPI.playing(self.view(for: node), resolve: resolve, reject: reject)
        }
    }
    
    @objc(currentAdBreak:resolver:rejecter:)
    func currentAdBreak(_ node: NSNumber, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        DispatchQueue.main.async {
            self.adsAPI.currentAdBreak(self.view(for: node), resolve: resolve, reject: reject)
        }
    }
    
    @objc(currentAds:resolver:rejecter:)
    func currentAds(_ node: NSNumber, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        DispatchQueue.main.async {
            self.adsAPI.currentAds(self.view(for: node), resolve: resolve, reject: reject)
        }
    }
    
    @objc(scheduledAdBreaks:resolver:rejecter:)
    func scheduledAdBreaks(_ node: NSNumber, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        DispatchQueue.main.async {
            self.adsAPI.scheduledAdBreaks(self.view(for: node), resolve: resolve, reject: reject)
        }
    }
    
    @objc(schedule:ad:)
    func schedule(_ node: NSNumber, adDict: NSDictionary) -> Void {
        DispatchQueue.main.async {
            self.adsAPI.schedule(self.view(for: node), adDict: adDict)
        }
    }
}
